Add force_destroy flag to ACLs and Dicts to allow deleting non-empty lists

bengesoff opened the pull request:
Updates service ACL and Dictionary blocks with extra protection to prevent the deletion of populated data that is dynamically added outside of a Terraform workflow.

Checks for non-empty ACL entries and Dictionary items before deleting ACLs and Dictionaries to reduce potential loss when updating either collection name.

Returns an error to the terraform user if `force_destroy` is not set on the ACL/Dictionary prior to updating.

bengesoff committed:
Add force_destroy to ACL block

bengesoff committed:
Improve ACL delete error message

bengesoff committed:
Apply versionless fix to dictionary too

bengesoff committed:
Update docs

Integralist reviewed (changes requested):
LGTM. But would be great to get a couple of tests in the relevant ACL/Dictionary test files to validate the new behaviour.

bengesoff committed:
Add integration tests for ACL/Dict force_destroy behaviour

Had to use a resource.TestCheckFunc with a side effect to add an
entry/item as part of a TestStep. This leads to a six part test where we
test renaming an empty list, add an entry, try renaming and expect an
error, then use force_destroy and retry renaming, expecting it to
succeed.

Integralist commented on fastly/block_fastly_service_v1_acl_test.go:
```diff
@@ -42,10 +43,18 @@ func TestResourceFastlyFlattenAcl(t *testing.T) {
+	// 3. Keep both ACLs the same and add an entry to one of them
+	// 4. Try to rename the ACLs, expect to fail with "list not empty error"
+	// 5. Without renaming the ACLs, set force_destroy=true to skip the deletion check
+	// 6. Try to rename the ACLs again, expect to succeed
```
❤️  the descriptive breakdown here. This has been lacking in the majority of the tests 👍🏻

bengesoff replied:
Thanks! Thought it was especially necessary given the longer-form test

Integralist commented on fastly/block_fastly_service_v1_acl_test.go:
```diff
@@ -55,31 +64,55 @@ func TestAccFastlyServiceV1_acl(t *testing.T) {
+			},
+			{
+				Config: testAccServiceV1Config_acl(name, aclNameUpdated),
+				Check:  testAccAddACLEntries(&acl),
```
```suggestion
				Check:  testAccAddACLEntries(&acl), // triggers side-effect of adding an ACL Entry
```

Integralist commented on fastly/block_fastly_service_v1_dictionary_test.go:
```diff
@@ -58,42 +68,47 @@ func TestAccFastlyServiceV1_dictionary(t *testing.T) {
+			},
+			{
+				Config:      testAccServiceV1Config_dictionary(name, dictName, backendName, domainName),
+				ExpectError: regexp.MustCompile(""),
```
Should we not get an error like we do with the ACL tests?

e.g.

```go
regexp.MustCompile("Cannot delete.*list is not empty.*")
```

bengesoff replied:
Oh that's a facepalm right there! Not sure where that's gone 😂

bengesoff replied:
Rerunning the test with this added back in 🤦🏻‍♂️

Integralist commented on fastly/block_fastly_service_v1_dictionary_test.go:
```diff
@@ -58,42 +68,47 @@ func TestAccFastlyServiceV1_dictionary(t *testing.T) {
+			},
+			{
+				Config: testAccServiceV1Config_dictionary(name, updatedDictName, backendName, domainName),
+				Check:  testAccAddDictionaryItems(&dictionary),
```
```suggestion
				Check:  testAccAddDictionaryItems(&dictionary), // triggers side-effect of adding a Dictionary Item
```

Integralist reviewed (commented):
LGTM @bengesoff just a couple of comments.

bengesoff committed:
Add extra side-effect comment to tests

Co-authored-by: Mark McDonnell <Integralist@users.noreply.github.com>

bengesoff committed:
Add missing regex pattern
